File: Data Structures & Algorithms/search-2d-matrix/submission-0.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {

        const rows = matrix.length
        const columns = matrix[0].length 
        let left = 0
        let right = ( rows * columns ) - 1

        while (left <= right){
            const mid = Math.floor((left + right) / 2)
            const row = Math.floor(mid / columns)
            const column = mid % columns 

            let current = matrix[row][column]

            if(current === target) return true 

            if(current > target){
                right = mid - 1
            }
            else{
                left = mid + 1
            }



        }
        return false 
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
class Solution:
    def isSubtree(self, root, subRoot):

        def isSameTree(p, q):
            
            if not p and not q :
                return True
            if not p or not q :
                return False 
            if p and q and p.val != q.val:
                return False 

            return (
                isSameTree(p.left, q.left)
                and
                isSameTree(p.right, q.right)
            )

        if not root:
            return False

        if isSameTree(root, subRoot):
            return True

        left = self.isSubtree(root.left, subRoot)
        right = self.isSubtree(root.right, subRoot)

        return left or right